File: RecordWorkouts/views/ToolbarBuilder.m
#import "ToolbarBuilder.h"

@implementation ToolbarBuilder {
    CGRect _frame;
    NSArray *defaultButtons;
    UIToolbar *defaultToolbar;
    
    NSMutableArray *customButtons;
}

@synthesize doneButton;
@synthesize spacer;

- (id) init: (CGRect)frame target:(id)target doneAction:(SEL)doneAction {
    self = [super init];
    _frame = frame;
    doneButton = [[UIBarButtonItem alloc] initWithTitle:@"Done" style:UIBarButtonItemStylePlain
                                              target:target action:doneAction];
    spacer = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil];
    defaultButtons = [NSArray arrayWithObjects:spacer, doneButton, nil];
    return self;
}

- (UIToolbar *) create {
    return [self createWith:customButtons];
}

- (UIToolbar *) createDefault {
    return [self createWith:defaultButtons];
}

- (UIToolbar *) createWith: (NSArray *)items {
    UIToolbar *toolBar= [[UIToolbar alloc] initWithFrame:_frame];
    [toolBar setTintColor:[UIColor grayColor]];
    [toolBar setItems:items];
    return toolBar;
}

- (void) addActionButton: (NSString *)title target: (id) target action: (SEL) action {
    UIBarButtonItem *button = [[UIBarButtonItem alloc] initWithTitle:title
       style:UIBarButtonItemStylePlain target:target action:action];
    [self addButton:button];
}

- (void) addButton: (UIBarButtonItem *)button {
    if (customButtons == nil) {
        customButtons = [[NSMutableArray alloc] init];
    }
    [customButtons addObject:button];
}

@end

File: workout-records/views/ToolbarBuilder.swift
import UIKit

class ToolbarBuilder {
    private var frame = CGRect.zero
    private var defaultButtons: [UIBarButtonItem] = []
    private var customButtons: [UIBarButtonItem] = []
    
    var doneButton: UIBarButtonItem
    var spacer: UIBarButtonItem
    
    init(_ frame: CGRect, target: Any?, doneAction: Selector) {
        self.frame = frame
        doneButton = UIBarButtonItem(title: "Done", style: .plain, target: target, action: doneAction)
        spacer = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
        defaultButtons = [spacer, doneButton]
    }
    
    func create() -> UIToolbar {
        return create(with: customButtons)
    }
    
    func createDefault() -> UIToolbar {
        return create(with: defaultButtons)
    }
    
    private func create(with items: [UIBarButtonItem]) -> UIToolbar {
        let toolBar = UIToolbar(frame: frame)
        toolBar.tintColor = UIColor.secondaryLabel
        toolBar.items = items
        return toolBar
    }
    
    func addActionButton(_ title: String, target: Any?, action: Selector) {
        let button = UIBarButtonItem(title: title, style: .plain, target: target, action: action)
        addButton(button)
    }
    
    func addButton(_ button: UIBarButtonItem) {
        customButtons.append(button)
    }
}
